File: mobile/lib/features/bullion_sales/domain/bullion_sale.dart
class BullionSale {
  const BullionSale({
    required this.id,
    required this.ref,
    required this.customerId,
    required this.customerName,
    required this.customerPhone,
    required this.customerArea,
    required this.salesmanId,
    required this.salesmanName,
    required this.weightGrams,
    required this.ratePerGram,
    required this.totalAmount,
    required this.transactionType,
    required this.goldPurity,
    required this.notes,
    required this.createdAt,
  });

  final String id;
  final String ref;
  final String customerId;
  final String customerName;
  final String customerPhone;
  final String customerArea;
  final String salesmanId;
  final String salesmanName;
  final double weightGrams;
  final double ratePerGram;
  final double totalAmount;
  final String transactionType;
  final String goldPurity;
  final String? notes;
  final DateTime createdAt;

  bool get isSale => transactionType == 'sale';
  bool get isReturn => transactionType == 'return';

  factory BullionSale.fromJson(Map<String, dynamic> json) {
    final customerObj = json['customer'];
    String customerId = '';
    String customerName = '';
    String customerPhone = '';
    String customerArea = '';

    if (customerObj is Map<String, dynamic>) {
      customerId = customerObj['id']?.toString() ?? customerObj['_id']?.toString() ?? '';
      customerName = customerObj['name']?.toString() ?? '';
      customerPhone = customerObj['phone']?.toString() ?? '';
      customerArea = customerObj['area']?.toString() ?? '';
    } else {
      customerId = json['customerId']?.toString() ?? '';
      customerName = json['customerName']?.toString() ?? '';
      customerPhone = json['customerPhone']?.toString() ?? '';
      customerArea = json['customerArea']?.toString() ?? '';
    }

    final salesmanObj = json['salesman'];
    String salesmanId = '';
    String salesmanName = '';
    if (salesmanObj is Map<String, dynamic>) {
      salesmanId = salesmanObj['id']?.toString() ?? salesmanObj['_id']?.toString() ?? '';
      salesmanName = salesmanObj['name']?.toString() ?? '';
    } else {
      salesmanId = json['salesmanId']?.toString() ?? '';
      salesmanName = json['salesmanName']?.toString() ?? '';
    }

    return BullionSale(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      ref: json['ref']?.toString() ?? '',
      customerId: customerId,
      customerName: customerName,
      customerPhone: customerPhone,
      customerArea: customerArea,
      salesmanId: salesmanId,
      salesmanName: salesmanName,
      weightGrams: (json['weightGrams'] as num?)?.toDouble() ?? 0.0,
      ratePerGram: (json['ratePerGram'] as num?)?.toDouble() ?? 0.0,
      totalAmount: (json['totalAmount'] as num?)?.toDouble() ?? 0.0,
      transactionType: json['transactionType']?.toString() ?? 'sale',
      goldPurity: json['goldPurity']?.toString() ?? '',
      notes: json['notes']?.toString(),
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }
}

class BullionSalesPage {
  const BullionSalesPage({
    required this.sales,
    required this.total,
    required this.page,
    required this.pages,
  });

  final List<BullionSale> sales;
  final int total;
  final int page;
  final int pages;

  factory BullionSalesPage.fromJson(Map<String, dynamic> json) {
    final dynamic listData = json['sales'];
    final List<dynamic> rawList = listData is List ? listData : [];
    return BullionSalesPage(
      sales: rawList
          .whereType<Map<String, dynamic>>()
          .map(BullionSale.fromJson)
          .toList(growable: false),
      total: (json['total'] as num?)?.toInt() ?? 0,
      page: (json['page'] as num?)?.toInt() ?? 1,
      pages: (json['pages'] as num?)?.toInt() ?? 1,
    );
  }
}

File: backend/src/models/BullionSale.js
import mongoose from 'mongoose'

const { Schema } = mongoose

const bullionSaleSchema = new Schema(
  {
    clientEntryId: {
      type: String,
      trim: true,
      default: null,
      index: true,
      unique: true,
      sparse: true,
    },
    customerId: {
      type: Schema.Types.ObjectId,
      ref: 'Customer',
      default: null,
      index: true,
    },
    customerName: {
      type: String,
      required: [true, 'Customer name is required'],
      trim: true,
    },
    customerPhone: {
      type: String,
      default: null,
      trim: true,
      index: true,
      validate: {
        validator: (value) => value === null || value === undefined || value === '' || /^\d{10}$/.test(String(value || '')),
        message: 'Customer phone must be exactly 10 digits',
      },
    },
    customerArea: {
      type: String,
      default: null,
      trim: true,
    },
    customerEmail: {
      type: String,
      default: null,
      trim: true,
      lowercase: true,
    },
    salesmanId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: [true, 'Salesman is required'],
      index: true,
    },
    salesmanName: {
      type: String,
      trim: true,
      default: null,
    },
    weightGrams: {
      type: Number,
      required: [true, 'Weight is required'],
      min: 0,
    },
    inputUnit: {
      type: String,
      enum: ['per_gram', 'per_kg'],
      default: 'per_gram',
    },
    ratePerGram: {
      type: Number,
      required: [true, 'Rate per gram is required'],
      min: 0,
    },
    totalAmount: {
      type: Number,
      required: [true, 'Total amount is required'],
      min: 0,
    },
    goldPurity: {
      type: String,
      trim: true,
      default: null,
    },
    transactionType: {
      type: String,
      enum: ['sale', 'return'],
      default: 'sale',
      index: true,
    },
    notes: {
      type: String,
      trim: true,
      default: null,
    },
    syncStatus: {
      type: String,
      enum: ['pending', 'synced', 'failed'],
      default: 'synced',
      index: true,
    },
    syncedAt: {
      type: Date,
      default: null,
    },
    syncError: {
      type: String,
      trim: true,
      default: null,
    },
    syncMeta: {
      type: Schema.Types.Mixed,
      default: null,
    },
  },
  {
    timestamps: true,
  }
)

bullionSaleSchema.index({ createdAt: -1 })
bullionSaleSchema.index({ salesmanId: 1, createdAt: -1 })
bullionSaleSchema.index({ customerId: 1, createdAt: -1 })
bullionSaleSchema.index({ transactionType: 1, createdAt: -1 })
bullionSaleSchema.index({ syncStatus: 1, createdAt: -1 })

bullionSaleSchema.methods.toSafeObject = function () {
  const obj = this.toObject()
  delete obj.__v
  return obj
}

export const BullionSale = mongoose.model('BullionSale', bullionSaleSchema)

